pep-9999.rst

```diff
@@ -0,0 +1,440 @@
+
+The proposed API takes a ``SupportsRead[bytes]``, while ``toml.load`` takes a
+``SupportsRead[str]`` and ``json.load`` takes ``SupportsRead[str | bytes]``.
+While slightly opinionated, this was changed in ``tomli`` v1.2 to a)
+ensure utf-8 is the encoding used, b) avoid incorrectly parsing single carriage
+returns as valid TOML due to universal newlines.
+
+Allowing users more control over formatting ``tomllib.dump[s]`` output
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As mentioned, TOML values can be represented in multiple ways, so inevitably,
+people will have strong opinions over how to do so.
+
+The ``toml`` library on PyPI supports this using custom subclasses of
+``toml.TomlEncoder``. There are a handful of instances of this that can be found
+on https://grep.app. However, the API to do this is not particularly clean.
+
+A non-exhaustive list of potential options users may want control over:
+
+* How to format strings
+* When to inline arrays or tables
+* How much to indent
+* Whether to reorder contents
+* Whether to use dotted keys
+
+In several cases, users could enforce TOML formatting by using an autoformatter
+of their choice at a later point.
+
+We acknowledge that supporting ``multiline_strings`` is something of an
+exception to this, if controversial we can err on the side of simplicity and
+remove it.
+
+Allowing users more control over ``tomllib.dump[s]`` serialisation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+It could be useful to add the equivalent of the ``default`` argument in ``json.dump``
+to allow users to specify how custom types should be serialised.
+
+The ``toml`` library on PyPI supports this using custom subclasses of
+``toml.TomlEncoder``. However, we could find two instances of using
+``toml.TomlEncoder`` to accomplish this kind of thing on https://grep.app, one
+of which was to add support to ``toml`` for dumping ``decimal.Decimal``.
+
+TOML is used more for configuration than serialisation of arbitrary data, so
+users are perhaps less likely to require custom serialisation than with say
+JSON. Support for this could be added in a backward compatible way.
+
+TODO: talk about output validation for ``dump[s]``
+
+Open Issues
+===========
+
+Package name
+------------
+
+Ideally, we would be able to use the ``toml`` package name. The ``toml`` package
+on PyPI is both widely used and not actively maintained.
+
+If the maintainer of ``toml`` resurfaced and was willing to give up the ``toml``
+name on PyPI, we could repurpose the PyPI package as a stdlib backport. However,
+this would still be breaking for users who have pinned current versions of the
+``toml`` package and have upgraded Python versions. The two API
+incompatibilities that most users of current ``toml`` would run into are a)
+different acceptable types to the first argument of ``toml.load``, b) use of
+``dump[s]`` if we choose not to include a write API.
```

Comment:
A third very major incompatibility IMO: `tomli.TOMLDecodeError` vs `toml.TomlDecodeError`

Reply:
Called this out in https://github.com/hauntsaninja/peps/pull/1/commits/7dc32cdfe1486028e5a234baa4df083cd3c94ee6